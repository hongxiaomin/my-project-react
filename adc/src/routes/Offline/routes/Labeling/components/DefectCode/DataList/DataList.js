/* eslint jsx-a11y/no-static-element-interactions: 0 */
import React from 'react'
import PropTypes from 'prop-types'
import WrapperFormattedMessage from 'components/WrapperFormattedMessage'
import { injectIntl } from 'react-intl'
import classes from './DataList.scss'

export class DataList extends React.Component {
  constructor(props) {
    super(props)

    this.input = null
    this.state = {
      text: '',
      select: false,
    }
  }

  onBlur = () => {
    this.setState({
      select: false,
    })
  }

  onDoubleClick = () => {
    this.setState({
      select: true,
    })
  }

  caretDownClick = () => {
    this.setState({
      select: !this.state.select,
    })
  }

  liClick = defect => () => {
    this.props.onChange(defect, () => {
      this.setState({
        select: true,
      })
    })
  }

  handleInputChange = (e) => {
    this.setState({
      text: this.input.value,
      select: true,
    })

    // this.props.onChange(value, () => {
    //   this.setState({
    //     select: true,
    //   })
    // })
  }

  cleanInput = () => {
    this.setState({
      text: '',
    })
    this.props.onChange('')
  }

  render() {
    let ul = ''
    const { selectedData, onFocus, onBlur, data, intl } = this.props
    const { text: stateText } = this.state
    const liList = []
    let inputIcon = (<span
      className={`pt-icon-standard pt-icon-small-cross ${classes['input-icon']}`}
      onClick={this.cleanInput}
    />)

    data.forEach((defect) => {
      if (!stateText || new RegExp(stateText.replace(/(\*|-)/g, '\\$1'), 'i').test(defect.defectCode)) {
        liList.push(<li className={`${classes.li}${selectedData.some(item => item.defectId === defect.defectId) ? ` ${classes.active}` : ''}`} key={defect.defectId} onClick={this.liClick(defect)}>{defect.defectCode}</li>)
      }
    })
    if (data.length && this.state.select) {
      if (!liList.length) liList.push(<li className={classes.li} key="no-data"><WrapperFormattedMessage id="labelingTool.dirTree.noData" /></li>)
      ul = (<ul className={classes.ul}>{liList}</ul>)
    }

    if (stateText === '') {
      inputIcon = (<span
        className={`pt-icon-standard pt-icon-caret-down ${classes['input-icon']}`}
        onClick={this.caretDownClick}
      />)
    }


    return (
      <div className={classes['DataList-container']}>
        {data.length && this.state.select
          ? <div className={classes['search-bg']} onClick={this.onBlur} />
          : ''
        }
        <div className={classes['input-div']}>
          <input
            ref={x => (this.input = x)}
            type="text"
            className="pt-input"
            value={this.state.text}
            onChange={this.handleInputChange}
            onFocus={onFocus}
            onBlur={onBlur}
            onDoubleClick={this.onDoubleClick}
            placeholder={intl.formatMessage({ id: 'pleaseSelect' })}
            style={{
              position: 'relative',
              width: '93%',
              border: 'none',
              height: '20px',
              lineHeight: '20px',
              padding: 0,
              boxShadow: 'none',
            }}
          />
          {inputIcon}
        </div>
        {ul}
      </div>
    )
  }
}

DataList.defaultProps = {
  onFocus: () => {},
  onBlur: () => {},
}

DataList.propTypes = {
  data: PropTypes.array.isRequired,
  onChange: PropTypes.func.isRequired,
  selectedData: PropTypes.array.isRequired,
  intl: PropTypes.object.isRequired,
  onFocus: PropTypes.func,
  onBlur: PropTypes.func,
}

export default injectIntl(DataList)
