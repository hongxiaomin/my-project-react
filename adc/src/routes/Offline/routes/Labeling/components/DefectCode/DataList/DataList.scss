.DataList-container {
  padding: 10px;
  position: relative;

  .ul {
    width: 94%;
    margin: 0 auto;
    border: 1px solid #ccc;
    overflow: hidden;
    overflow-y: auto;
    max-height: 200px;
    padding: 0;
    position: absolute;
    background: #fff;
    z-index: 99;
    left: 50%;
    margin-left: -47%;
    border-radius: 5px;
  }
  .li{
    list-style: none;
    text-align: left;
    background: none;
    padding: 5px 10px;
    word-wrap: break-word;

    &:hover, &.active{
      background: rgba(98, 189, 245, 0.77);
      cursor: default;
    }
  }
  .input-div{
    position: relative;
    width: 100%;
    box-shadow: 0 0 0 0 rgba(19, 124, 189, 0), 0 0 0 0 rgba(19, 124, 189, 0), inset 0 0 0 1px rgba(16, 22, 26, 0.15), inset 0 1px 1px rgba(16, 22, 26, 0.2);
    padding: 5px;
    border-radius: 5px;
  }
  .input-icon{
    position: absolute;
    background: none;
    cursor: pointer;
    right: 5px;

    &:active,
    &:hover{
      background: #ccc !important;
    }


  }
  .input-icon-cross {
    background: #fff;
  }
  .search-bg {
    background: rgba(0, 0, 0, 0);
    position: fixed;
    width: 100%;
    height: 100%;
    left: 0;
    top: 0;
    z-index: 19;
  }
}
