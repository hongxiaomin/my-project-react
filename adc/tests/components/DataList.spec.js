import React from 'react'
import { shallow, mount } from 'enzyme'
import _ from 'lodash'
import { DataList } from 'routes/Offline/routes/Labeling/components/DefectCode/DataList/DataList'

describe('(Component) DataList', () => {
  const data = []
  const onChange = jest.fn()
  it('render normally', () => {
    const wrapper = shallow(
      <DataList
        data={data}
        onChange={onChange}
      />,
    )
    // const nativeEvent = { nativeEvent: { stopImmediatePropagation: _.noop } }
    expect(wrapper).toMatchSnapshot()
  })
  it('span click', () => {
    const wrapper = shallow(
      <DataList
        data={data}
        onChange={onChange}
      />,
    )
    wrapper.find('span').at(0).simulate('click')
    expect(wrapper.state('select')).toEqual(true)
    wrapper.find('span').at(1).simulate('click')
    expect(wrapper.state('text')).toEqual('')
  })
})

