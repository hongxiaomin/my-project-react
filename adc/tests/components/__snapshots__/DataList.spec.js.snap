// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(Component) DataList render normally 1`] = `
<div
  className="DataList-container"
>
  <div
    className="input-div"
  >
    <input
      className="pt-input"
      onBlur={[Function]}
      onChange={[Function]}
      onDoubleClick={[Function]}
      onFocus={[Function]}
      placeholder="Please select"
      style={
        Object {
          "border": "none",
          "boxShadow": "none",
          "height": "20px",
          "lineHeight": "20px",
          "padding": 0,
          "position": "relative",
          "width": "93%",
        }
      }
      type="text"
      value=""
    />
    <span
      className="pt-icon-standard pt-icon-caret-down input-icon"
      onClick={[Function]}
    />
  </div>
</div>
`;
